src/thermo/PengRobinson.cpp

```diff
@@ -0,0 +1,955 @@
+}
+
+double PengRobinson::cv_mole() const
+{
+    _updateReferenceStateThermo();
+    double T = temperature();
+    pressureDerivatives();
+    return (cp_mole() + T* dpdT_* dpdT_ / dpdV_);
+}
+
+double PengRobinson::pressure() const
+{
+    _updateReferenceStateThermo();
+    //  Get a copy of the private variables stored in the State object
+    double T = temperature();
+    double mv = molarVolume();
+    double denom = mv * mv + 2 * mv * m_b_current - m_b_current * m_b_current;
+    double pp = GasConstant * T / (mv - m_b_current) - m_aAlpha_current / denom;
+    return pp;
+}
+
+double PengRobinson::standardConcentration(size_t k) const
+{
+    getStandardVolumes(m_tmpV.data());
+    return 1.0 / m_tmpV[k];
+}
+
+void PengRobinson::getActivityCoefficients(double* ac) const
+{
+    double mv = molarVolume();
+    //double T = temperature();
+    double vpb2 = mv + (1 + M_SQRT2)*m_b_current;
+    double vmb2 = mv + (1 - M_SQRT2)*m_b_current;
+    double vmb = mv - m_b_current;
+    double pres = pressure();
+
+    for (size_t k = 0; k < m_kk; k++) {
+        m_pp[k] = 0.0;
+        for (size_t i = 0; i < m_kk; i++) {
+            size_t counter = k + m_kk*i;
+            m_pp[k] += moleFractions_[i] * aAlpha_vec_Curr_[counter];
+        }
+    }
+    double num = 0;
+    double den = 2 * M_SQRT2 * m_b_current * m_b_current;
+    double den2 = m_b_current*(mv * mv + 2 * mv * m_b_current - m_b_current * m_b_current);
+    double RTkelvin = RT();
+    for (size_t k = 0; k < m_kk; k++) {
+        num = 2 * m_b_current * m_pp[k] - m_aAlpha_current* b_vec_Curr_[k];
+        ac[k] = (-RTkelvin *log(pres*mv/ RTkelvin) + RTkelvin * log(mv / vmb)
+                 + RTkelvin * b_vec_Curr_[k] / vmb
+                 - (num /den) * log(vpb2/vmb2)
+                 - m_aAlpha_current* b_vec_Curr_[k] * mv/den2
+                );
+    }
+    for (size_t k = 0; k < m_kk; k++) {
+        ac[k] = exp(ac[k]/ RTkelvin);
+    }
+}
+
+// ---- Partial Molar Properties of the Solution -----------------
+
+void PengRobinson::getChemPotentials_RT(double* muRT) const
+{
+    getChemPotentials(muRT);
+    double RTkelvin = RT();
+    for (size_t k = 0; k < m_kk; k++) {
+        muRT[k] *= 1.0 / RTkelvin;
+    }
+}
+
+void PengRobinson::getChemPotentials(double* mu) const
+{
+    getGibbs_ref(mu);
+    double RTkelvin = RT();
+    for (size_t k = 0; k < m_kk; k++) {
+        double xx = std::max(SmallNumber, moleFraction(k));
+        mu[k] += RTkelvin *(log(xx));
+    }
+
+    double mv = molarVolume();
+    double vmb = mv - m_b_current;
+    double vpb2 = mv + (1 + M_SQRT2)*m_b_current;
+    double vmb2 = mv + (1 - M_SQRT2)*m_b_current;
+
+    for (size_t k = 0; k < m_kk; k++) {
+        m_pp[k] = 0.0;
+        for (size_t i = 0; i < m_kk; i++) {
+            size_t counter = k + m_kk*i;
+            m_pp[k] += moleFractions_[i] * aAlpha_vec_Curr_[counter];
+        }
+    }
+    double pres = pressure();
+    double refP = refPressure();
+    double num = 0;
```

Comment:
delclare `num` inside the loop, where it is really initialized.

Reply:
num is declared before the loop, just to avoid repetitive declarations inside the loop.

Reply:
Yes. I was mainly thinking from the point of better performance, but as you said that won't be an issue. The variable is now declared inside the loop.